### PR TITLE
fix(Link): Set focus back to the body element

### LINKED ISSUE #7681: Routing isn't screen-reader accessible?
# Bug report

I was browsing my next.js app with voiceover on, and noticed that client-side route changes aren't announced or even noticed by the screen reader. This is kind of alarming, since I've been using it in production on the assumption that the routing was accessible (my bad, was confusing with gatsby's use of reach-router).

I don't see many other references to "accessible routing" or "screen readers" in the issue history or documentation.

## To Reproduce

With voiceover (perhaps other screen readers too) on, click a next/link.

## Expected behavior

Focus should move to the top of the new page, and the new page title should be announced by the screen reader, like with a server-side route change, when clicking a next/link.

**I'm not a regular screen reader user, so it's possible I'm missing something and the current implementation is actually fine, but I think the point is that the server behaviour and client-rendered behaviours should be the same?**

## System information

- macOS
- Chrome and Firefox
- Version of Next.js: 8.1

### PR DESCRIPTION
**Pull request by connor-baer**
Follow up to #3545. 
Fixes #7681. 

**TL;DR**: `document.activeElement.blur()` reliably moves the keyboard focus back to the `body` element.

## Approach & changes

Turns out [my solution](https://github.com/zeit/next.js/pull/3545) from 2 years ago didn't do anything. The focus moved back to the `body` element automatically if the link element was removed from the DOM, which often happens when navigating to a new page.

However, if the link persisted on the new page, e.g. because it is inside `_app.js`, then the focus stays on the link element. The body element cannot be focused with `.focus()`, because it's simply not a keyboard-focusable element. We could fix that with `tabindex="-1"`, but that would cause other issues. 

`document.activeElement.blur()` unfocuses the current (link) element and the focus then moves back to the default focus element – the `body` element.

### CODE REVIEW

**Comment by connor-baer** [comment]
About the TypeScript issues: `document.activeElement` is an [`Element`](https://developer.mozilla.org/en-US/docs/Web/API/Element) type, not [`HTMLElement`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement), so it doesn't technically support `blur()`. Does anyone have an idea how to get around this?

File: packages/next/client/link.tsx [hunk 1]
```diff
@@ -180,7 +180,9 @@ class Link extends Component<LinkProps> {
         if (!success) return
         if (scroll) {
           window.scrollTo(0, 0)
-          document.body.focus()
+          // set focus back to the body element
+          const activeElement = document.activeElement as HTMLElement
+          activeElement.blur()
```
Comment (ijjk) [inline packages/next/client/link.tsx]:
```suggestion
          if (activeElement) {
            activeElement.blur()
          }
```

**Comment by ijjk** [comment]
Hi thanks for the PR, can you add a test for this making sure the correct `activeElement` is focused after a client transition? This will help make sure we maintain the correct behavior moving forward.

File: packages/next/client/link.tsx (hunk 1 above)
Comment (Timer) [inline packages/next/client/link.tsx]: We should check if this exists right? What if it's an element that doesn't support `blur()`, is that possible?

Comment (connor-baer) [reply packages/next/client/link.tsx]: AFAIK, there's always one active element, but of course, it doesn't hurt to add a check 👍

`blur()` is a standard method on all HTMLElements: https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/blur

**Comment by connor-baer** [comment]
@ijjk I'll add the tests asap. Currently a bit swamped at work 😬

File: packages/next/client/link.tsx (hunk 1 above)
Comment (Timer) [reply packages/next/client/link.tsx]: Sure, but you said the type is actually `Element` not `HTMLElement` in a previous comment: https://github.com/zeit/next.js/pull/7693#issuecomment-506532399.

**Review by ijjk — APPROVED** [review]

**Comment by ijjk** [comment]
@connor-baer I went ahead and added the tests and added an additional check to the blurring logic since you're busy and we want to make sure this is handled. Thanks for the taking the time to submit the PR!

**Review by Timer — APPROVED** [review]

**Comment by connor-baer** [comment]
Hey @ijjk, thanks for adding the tests and finishing the PR 🙌

I saw that you reverted the changes in #7753 — I assume because it broke something? Could you please elaborate on the issue? I have more free time this week and would love to fix this.
